azurerm/resource_arm_kubernetes_cluster.go

```diff
@@ -678,21 +592,47 @@ func resourceArmKubernetesClusterCreate(d *schema.ResourceData, meta interface{}
 	dnsPrefix := d.Get("dns_prefix").(string)
 	kubernetesVersion := d.Get("kubernetes_version").(string)
 
-	linuxProfile := expandKubernetesClusterLinuxProfile(d)
-	agentProfiles, err := expandKubernetesClusterAgentPoolProfiles(d)
+	linuxProfileRaw := d.Get("linux_profile").([]interface{})
+	linuxProfile := expandKubernetesClusterLinuxProfile(linuxProfileRaw)
+
+	agentProfiles, err := containers.ExpandDefaultNodePool(d)
 	if err != nil {
-		return err
+		return fmt.Errorf("Error expanding `default_node_pool`: %+v", err)
 	}
-	windowsProfile := expandKubernetesClusterWindowsProfile(d)
-	servicePrincipalProfile := expandAzureRmKubernetesClusterServicePrincipal(d)
-	networkProfile := expandKubernetesClusterNetworkProfile(d)
-	addonProfiles := expandKubernetesClusterAddonProfiles(d)
 
-	t := d.Get("tags").(map[string]interface{})
+	// TODO: remove me in 2.0
+	if agentProfiles == nil {
+		agentProfilesRaw := d.Get("agent_pool_profile").([]interface{})
+		agentProfilesLegacy, err := expandKubernetesClusterAgentPoolProfiles(agentProfilesRaw, true)
+		if err != nil {
+			return err
+		}
+
+		agentProfiles = &agentProfilesLegacy
+	}
+
+	addOnProfilesRaw := d.Get("addon_profile").([]interface{})
+	addonProfiles := containers.ExpandKubernetesAddOnProfiles(addOnProfilesRaw)
+
+	networkProfileRaw := d.Get("network_profile").([]interface{})
+	networkProfile := expandKubernetesClusterNetworkProfile(networkProfileRaw)
 
 	rbacRaw := d.Get("role_based_access_control").([]interface{})
 	rbacEnabled, azureADProfile := expandKubernetesClusterRoleBasedAccessControl(rbacRaw, tenantId)
 
+	// since the Create and Update use separate methods, there's no point extracting this out
+	servicePrincipalProfileRaw := d.Get("service_principal").([]interface{})
+	servicePrincipalProfileVal := servicePrincipalProfileRaw[0].(map[string]interface{})
```

Comment:
Should we check for length > 0 and that it's not nil before casting this out?

Reply:
this is required - so fine for now - when this becomes optional in the future we can handle that then 👍